src/Workspaces/Core/Portable/Workspace/Workspace_EventsLegacy.cs

```diff
@@ -76,65 +78,63 @@ public event EventHandler<TextDocumentEventArgs> TextDocumentClosed
             {
-                (var adviseCount, disposer) = adviseCountAndDisposer;
-
                 // If we already have a handler for this event type, update the map with the new advise count
-                _disposableEventHandlers[(eventHandler, eventType)] = (adviseCount + 1, disposer);
+                _disposableEventHandlers[key] = (adviseCountAndDisposer.adviseCount + 1, adviseCountAndDisposer.disposer);
             }
-        }
-
-        if (disposer == null)
-        {
-            disposer = RegisterHandler(eventType, (Action<EventArgs>)Handler, WorkspaceEventOptions.RequiresMainThreadOptions);
-
-            lock (_gate)
+            else
             {
-                _disposableEventHandlers[(eventHandler, eventType)] = (AdviseCount: 1, disposer);
+                // Safe to call RegisterHandler inside the lock as it doesn't invoke code outside the workspace event map code.
+                var disposer = RegisterHandler(eventType, (Action<EventArgs>)Handler, WorkspaceEventOptions.RequiresMainThreadOptions);
+                _disposableEventHandlers[key] = (adviseCount: 1, disposer);
             }
         }
 
+        return;
+
         void Handler(EventArgs arg)
             => eventHandler(sender: this, (TEventArgs)arg);
     }
 
-    private void RemoveEventHandler<TEventArgs>(EventHandler<TEventArgs> eventHandler, WorkspaceEventType eventType)
+    private void RemoveLegacyEventHandler<TEventArgs>(EventHandler<TEventArgs> eventHandler, WorkspaceEventType eventType)
         where TEventArgs : EventArgs
     {
-        var existingAdviseCount = 0;
         IDisposable? disposer = null;
 
-        lock (_gate)
+        lock (_legacyWorkspaceEventsGate)
         {
             if (_disposableEventHandlers.TryGetValue((eventHandler, eventType), out var adviseCountAndDisposer))
             {
                 // If we already have a handler for this event type, just increment the advise count
                 // and return.
-                (existingAdviseCount, disposer) = adviseCountAndDisposer;
-                if (existingAdviseCount == 1)
+                if (adviseCountAndDisposer.adviseCount == 1)
+                {
+                    disposer = adviseCountAndDisposer.disposer;
                     _disposableEventHandlers.Remove((eventHandler, eventType));
+                }
                 else
-                    _disposableEventHandlers[(eventHandler, eventType)] = (existingAdviseCount - 1, disposer);
+                {
+                    _disposableEventHandlers[(eventHandler, eventType)] = (adviseCountAndDisposer.adviseCount - 1, adviseCountAndDisposer.disposer);
+                }
             }
         }
 
-        Debug.Assert(existingAdviseCount > 0, $"Event handler for event type {eventType} was not registered.");
+        Debug.Assert(disposer != null, $"Event handler for event type {eventType} was not registered.");
```

Comment:
This assert isn't correct since a decrement of the advise count will leave it assigned to null, but the assert message doesn't jive with that. Probably easiest to just remove the assert.